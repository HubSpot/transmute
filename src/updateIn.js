import curry from './curry';
import _getIn from './internal/_getIn';
import _setIn from './internal/_setIn';

function updateIn(keyPath, updater, subject) {
  const value = _getIn(keyPath, subject);
  return _setIn(keyPath, updater(value), subject);
}

/**
 * Apply `updater` to the value at `keyPath`.
 *
 * @example
 * const incrementUserCount = updateIn(['users', 'count'], n => n + 1);
 * incrementUserCount({users: {count: 1}});
 * // returns {users: {count: 2}}
 *
 * @example <caption>Unset keyPaths will be set based on the most recent type.</caption>
 * const incrementUserCount = updateIn(['users', 'count'], (n = 0) => n + 1);
 * incrementUserCount({});
 * // returns {users: {count: 1}}
 *
 * incrementUserCount(Map());
 * // returns Map { users => Map { count => 1 } }
 *
 * @param  {Array<any>|Iterable<any>} keyPath the location where `updater` should be applied.
 * @param  {Function} updater the tranformation to apply.
 * @param  {Array|Iterable|Object} subject the thing to update.
 * @return {Array|Iterable|Object}
 */
export default curry(updateIn);
