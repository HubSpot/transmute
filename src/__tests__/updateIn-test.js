import { Map } from 'immutable';
import updateIn from '../updateIn';

describe('transmute/updateIn', () => {
  const incTwo = updateIn(['one', 'two'], n => n + 1);

  it('updates a nested Map', () => {
    expect(
      incTwo(
        Map({
          one: Map({
            two: 2,
          }),
        })
      )
    ).toMatchSnapshot();
  });

  it('updates a nested Object', () => {
    expect(
      incTwo({
        one: {
          two: 2,
        },
      })
    ).toMatchSnapshot();
  });
});
