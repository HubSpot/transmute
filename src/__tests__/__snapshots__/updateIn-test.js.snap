// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/updateIn updates a nested Map 1`] = `
Immutable.Map {
  "one": Immutable.Map {
    "two": 3,
  },
}
`;

exports[`transmute/updateIn updates a nested Object 1`] = `
Object {
  "one": Object {
    "two": 3,
  },
}
`;
